core/src/main/java/org/elasticsearch/common/inject/SpawnShittyModules.java

```diff
@@ -20,9 +20,19 @@
 package org.elasticsearch.common.inject;
 
 /**
+ * This interface can be added to a Module to spawn sub modules. DO NOT USE.
  *
+ * This is fundamentally broken.
+ * <ul>
+ * <li>If you have a plugin with multiple modules, return all the modules at once.</li>
+ * <li>If you are trying to make the implementation of a module "pluggable", don't do it.
+ * This is not extendable because custom implementations (using onModule) cannot be
+ * registered before spawnModules() is called.</li>
+ * </ul>
+ * @deprecated Create modules directly
  */
-public interface SpawnModules {
+@Deprecated
```

Comment:
> if removing the crappy api introduces another crappy thing (api/design/etc..)

But that's not the case here. Flattening modules is a good thing. It will make dependencies of services/components much less "spaghetti". Obviously I can't do flattening/cleanup in this PR. But the deprecation is here to warn against adding more of this craziness.

> The problem here is that there needs to be a well defined function for replacing modules, on the container level..

I disagree. There should be no need to replace a module, there is no need to make every part of Elasticsearch pluggable. We should have very specific pieces that are overrideable. Using the DiscoveryModule as an example, we have "local" and "zen" discovery types that are "builtin" to ES. These are always available, and zen has a number of associated classes that are bound in discovery module. However, if a plugin wants to add a new discovery type, the Discovery class is the only thing they are plugging in, and that is registerable with the DiscoveryModule. If they require extra bindings, they can do so in their own module, which is added as part of their plugin. All of our extension points should work this way.